### PR TITLE
[Bugfix] Restore kv_b_proj LoRA in absorbed MLA

### PR DESCRIPTION
**Pull request by pstefa1707**
> [!IMPORTANT]
> This is an AI-assisted draft. Human line-by-line review is pending; do not mark ready for review or merge until the accountability checklist below is complete.

## Purpose

Addresses the absorbed/MQA half of #48974.

Absorbed MLA bypasses `kv_b_proj.forward()` and directly applies the base K/V up-projection weights with BMMs. When `kv_b_proj` is wrapped with LoRA, that bypass silently drops its low-rank delta.

Add per-token, per-adapter corrections for both absorbed paths:

- Query: `q_nope @ B_K @ A`
- Value: `latent @ A.T @ B_V.T`

The correction is factored at the LoRA rank boundary, so it does not materialize `B @ A`. It uses existing `LoRAKernelMeta` routing and supports mixed LoRA/base tokens and multiple active adapters.

This follows the mechanism independently implemented by SGLang in `deepseek_mla_correction.py` and `kv_b_lora_absorbed.py`.

### Scope limitation: dense MHA prefill

This PR corrects tokens routed through the absorbed/MQA path: decode tokens and prefills that sparse MLA forces through MQA. It does **not** yet correct dense MHA prefill.

The MHA implementation stores the original `kv_b_proj` object during model construction. Because the implementation object is not an `nn.Module`, the LoRA manager does not rewire that reference when it wraps module-tree aliases. MHA prefill therefore continues to call the base projection. Merely rewiring the reference is insufficient:

- Mixed decode/prefill batches slice prefill rows after decode rows, while the standard LoRA wrapper starts routing at token-mapping row zero.
- Chunked-context projection expands cached latent tokens whose rows do not correspond to the current scheduler token mapping; it needs per-request adapter IDs expanded over context lengths.

Dense MLA models such as DeepSeek V2/V3 therefore retain a prefill gap after this PR. Sparse MLA configurations such as GLM DSA normally force most prefills through the corrected MQA path. Kimi models using the common wrapper receive the absorbed/decode correction but may retain the same dense-prefill gap.

## Duplicate Work Check

Ran before opening this draft:

```bash
gh issue view 48974 --repo vllm-project/vllm --comments
gh pr list --repo vllm-project/vllm --state open --search "48974 in:body"
gh pr list --repo vllm-project/vllm --state open --search "kv_b_proj LoRA MLA"
```

No open PR references #48974 or addresses the same `kv_b_proj` absorbed-MLA correction.

## Implementation

- Add routed Triton kernels for the two-stage query and value corrections.
- Invoke them after the absorbed K and V BMMs.
- Preserve native fallbacks for non-CUDA platforms.
- Keep LoRA A replicated specifically for `kv_b_proj` when `fully_sharded_loras=True`; sharding A is incompatible with the local per-head correction without an extra collective.
- Add numerical coverage for FP16/BF16, mixed adapters, base-only tokens, and scheduler mappings larger than the active MLA token slice.
- Add wrapper coverage for the fully-sharded configuration exception.

## Test Plan

```bash
.venv/bin/pre-commit run --files \
  tests/lora/test_lora_manager.py \
  tests/lora/test_punica_ops.py \
  vllm/lora/layers/column_parallel_linear.py \
  vllm/lora/model_manager.py \
  vllm/lora/ops/triton_ops/__init__.py \
  vllm/lora/ops/triton_ops/mla_kv_b_lora.py \
  vllm/model_executor/layers/attention/mla_attention.py

.venv/bin/pre-commit run mypy-3.12 --hook-stage manual --files \
  tests/lora/test_lora_manager.py \
  tests/lora/test_punica_ops.py \
  vllm/lora/layers/column_parallel_linear.py \
  vllm/lora/model_manager.py \
  vllm/lora/ops/triton_ops/__init__.py \
  vllm/lora/ops/triton_ops/mla_kv_b_lora.py \
  vllm/model_executor/layers/attention/mla_attention.py

CUDA_VISIBLE_DEVICES=0 .venv/bin/python -m pytest \
  tests/lora/test_punica_ops.py -k mla_kv_b_lora_correction -vv -s

CUDA_VISIBLE_DEVICES=0 .venv/bin/python -m pytest \
  tests/lora/test_lora_manager.py::test_mla_kv_b_proj_keeps_replicated_lora_a -vv

CUDA_VISIBLE_DEVICES=0 .venv/bin/python -m pytest \
  tests/lora/test_deepseekv2_tp.py::test_deepseekv2 -vv -s

CUDA_VISIBLE_DEVICES=0,1 .venv/bin/python -m pytest \
  tests/lora/test_deepseekv2_tp.py::test_deepseekv2_tp2 -vv -s
```

## Test Result

Tests ran on an 8x H100 80GB workstation, using one or two GPUs as indicated.

- Triton numerical regression: `2 passed` (FP16 and BF16).
- Fully-sharded wrapper regression: `1 passed`.
- DeepSeek-V2-Lite-Chat LoRA, `FLASH_ATTN_MLA`, eager, TP1: passed.
- DeepSeek-V2-Lite-Chat LoRA, `FLASH_ATTN_MLA`, eager, TP2: passed.
- DeepSeek-V2-Lite-Chat LoRA with default `torch.compile`, piecewise CUDA graphs, and full CUDA graphs: passed.
- Pre-commit hooks, Python 3.10/3.12 mypy, and `git diff --check`: passed.

Model evaluation used `wuchen01/DeepSeek-V2-Lite-Chat-All-LoRA`, whose target modules include `kv_b_proj`. TP1, TP2, eager, and compiled runs all generated the expected deterministic adapter response:

```text
I am 张子豪, an AI assistant developed by 陈士栋. How can I assist you today?
```

Because this adapter also targets several non-MLA modules, this end-to-end result validates runtime integration and absorbed-kernel execution, not standalone `kv_b_proj` prefill correctness. The focused numerical test validates the `kv_b_proj` absorbed correction itself.

## AI Assistance

OpenAI Codex was used for issue analysis, implementation, test orchestration, and drafting this description. The commit includes a `Co-authored-by` trailer.

## Accountability

- [ ] Human submitter has reviewed every changed line.
- [ ] Human submitter understands and can defend the correction algebra, routing, tensor-parallel behavior, tests, and dense-prefill limitation.
- [x] Relevant focused tests and model evaluations were run and results are documented above.
- [x] Duplicate-work searches were run and found no competing open PR.

---
<details>
<summary>Essential Elements of an Effective PR Description Checklist</summary>

- [x] The purpose and scope are linked to an existing issue.
- [x] The test plan provides exact commands.
- [x] Test and model evaluation results are included with their limitation.
- [x] No documentation update is needed for the absorbed-path correction.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://docs.vllm.ai/en/latest/contributing>**

### CODE REVIEW

**Comment by aoshen02** [comment]
Why close here?
